File: SortingAlgorithms/java/QuickSort.java
/******************************************************************************
QUICK SORT (Java)

Test case - 1:

Input:
5
10 22 78 91 38

Ouput:
10 22 38 78 91
-------------------------------------------
Test case - 2:

Input:
10
9 76 229 20 63 92 38 19 71 91

Ouput:
9 19 20 38 63 71 76 91 92 229
*******************************************************************************/
import java.util.Scanner;

public class QuickSort {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Read the size of the array and the array from console
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        // Call the quick sort function
        quickSort(arr, 0, n - 1);

        // Print the sorted array
        for (int i = 0; i < n; i++)
            System.out.print(arr[i] + " ");

        sc.close();
    }

    /**
     * The method that sorts the array using QuickSort algorithm
     * 
     * @param arr  The array that is to be sorted
     * @param low  Starting index
     * @param high Ending index
     */
    private static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // PI is partitioning index, arr[pi] is now at right place
            int pi = partition(arr, low, high);

            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    /**
     * This function takes last element as pivot an places the pivot element at its
     * right place. It also places smaller elements on the left of this pivot
     * element It also places greater elements on the right of this pivot element
     * 
     * @param arr
     * @param low
     * @param high
     * @return
     */
    private static int partition(int[] arr, int low, int high) {

        // Pivot is the element that needs to be placed in the right position
        int pivot = arr[high];

        // counter is the index of smaller element
        int counter = (low - 1);
        for (int i = low; i < high; i++) {

            // If current element is smaller than pivot
            if (arr[i] < pivot) {
                // Increment index of smaller element
                counter++;

                // Swap the smaller element and current element
                int temp = arr[counter];
                arr[counter] = arr[i];
                arr[i] = temp;
            }
        }

        // Swap the last element
        int temp = arr[counter + 1];
        arr[counter + 1] = arr[high];
        arr[high] = temp;

        // return the index of element sorted
        return counter + 1;
    }
}

File: SortingAlgorithms/Kotlin/QuickSort.kt
class QuickSort {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val array = arrayOf(90, 56, 101, 45, 65, 99, 67, 89, 34, 23)
            println("Before Sorted")
            for (element in array) {
                print("$element ")
            }
            println()
            quickSort(array, 0, 9)
            println("Sorted Array")
            for (element in array) {
                print("$element ")
            }
        }

        private fun partition(a: Array<Int>, beg: Int, end: Int): Int {    //int a[], int beg, int end
            var left: Int = 0
            var right: Int
            var temp: Int
            var loc: Int
            var flag: Int
            loc = beg.also({ left = it })
            right = end
            flag = 0
            while (flag != 1) {
                while (a.get(loc) <= a.get(right) && loc != right) right--
                if (loc == right) flag = 1 else if (a.get(loc) > a.get(right)) {
                    temp = a.get(loc)
                    a[loc] = a.get(right)
                    a[right] = temp
                    loc = right
                }
                if (flag != 1) {
                    while (a.get(loc) >= a[left] && loc != left) left++
                    if (loc == left) flag = 1 else if (a.get(loc) < a.get(left)) {
                        temp = a.get(loc)
                        a[loc] = a.get(left)
                        a[left] = temp
                        loc = left
                    }
                }
            }
            return loc
        }

        private fun quickSort(a: Array<Int>, beg: Int, end: Int) {
            var loc: Int
            if (beg < end) {
                loc = partition(a, beg, end)
                quickSort(a, beg, loc - 1)
                quickSort(a, loc + 1, end)
            }
        }
    }
}
